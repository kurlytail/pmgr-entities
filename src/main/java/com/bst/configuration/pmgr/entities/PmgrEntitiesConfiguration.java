package com.bst.configuration.pmgr.entities;

import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
import org.springframework.boot.autoconfigure.domain.EntityScan;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.EnableAspectJAutoProxy;
import org.springframework.data.jpa.repository.config.EnableJpaAuditing;
import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
import org.springframework.transaction.annotation.EnableTransactionManagement;

@Configuration
@EntityScan({"com.bst.pmgr.entities.generic", "com.bst.pmgr.entities"})
@EnableJpaRepositories("com.bst.pmgr.entities.repositories")
//@ImportResource({"classpath:Aspect.xml"})
@EnableTransactionManagement
@EnableJpaAuditing
@EnableAspectJAutoProxy
@EnableAutoConfiguration
public class PmgrEntitiesConfiguration {

}
