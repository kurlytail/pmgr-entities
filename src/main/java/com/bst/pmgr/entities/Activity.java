package com.bst.pmgr.entities;

import javax.persistence.DiscriminatorColumn;
import javax.persistence.Entity;
import javax.persistence.EntityListeners;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.InheritanceType;
import javax.persistence.ManyToOne;

import com.bst.pmgr.entities.audit.ActivityAudit;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

@EntityListeners(ActivityAudit.class)
@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "activityType")
@ToString
@Getter
@EqualsAndHashCode
public class Activity {
    @Id
    @GeneratedValue(strategy=GenerationType.AUTO)
    private Long id;

	@Setter
	private String name;
	
	@Setter
	private String metaName;
    
	@ManyToOne
	private Work work;
}
