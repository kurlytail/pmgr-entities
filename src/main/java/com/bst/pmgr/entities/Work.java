package com.bst.pmgr.entities;

import java.util.ArrayList;
import java.util.List;

import javax.persistence.Column;
import javax.persistence.DiscriminatorColumn;
import javax.persistence.Entity;
import javax.persistence.EntityListeners;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.InheritanceType;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.validation.constraints.NotNull;

import com.bst.pmgr.entities.audit.WorkAudit;

@EntityListeners(WorkAudit.class)
@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "workType")
public class Work {
	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	private Long id;

	@Column(nullable = false, updatable = false)
	private String name;

	public Long getId() {
		return id;
	}

	public void setId(Long id) {
		this.id = id;
	}

	public void setName(String name) {
		this.name = name;
	}

	public String getName() {
		return name;
	}

	public List<Activity> getActivities() {
		return activities;
	}

	public List<Document> getDocuments() {
		return documents;
	}

	public List<Tool> getTools() {
		return tools;
	}

	public List<Process> getProcesses() {
		return processes;
	}

	public List<ProcessGroup> getProcessGroups() {
		return processGroups;
	}

	public PmgrPerson getCreatedBy() {
		return createdBy;
	}

	public void setCreatedBy(PmgrPerson createdBy) {
		this.createdBy = createdBy;
	}

	@OneToMany(mappedBy = "work")
	private List<Activity> activities = new ArrayList<Activity>();

	@OneToMany(mappedBy = "work")
	private List<Document> documents = new ArrayList<Document>();

	@OneToMany(mappedBy = "work")
	private List<Tool> tools = new ArrayList<Tool>();

	@OneToMany(mappedBy = "work")
	private List<Process> processes = new ArrayList<Process>();

	@OneToMany(mappedBy = "work")
	private List<ProcessGroup> processGroups = new ArrayList<ProcessGroup>();

	@NotNull
	@ManyToOne
	private PmgrPerson createdBy;
}
