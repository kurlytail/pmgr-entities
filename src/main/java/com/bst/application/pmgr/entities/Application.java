package com.bst.application.pmgr.entities;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.web.servlet.support.SpringBootServletInitializer;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.transaction.annotation.EnableTransactionManagement;

import com.bst.configuration.pmgr.entities.PmgrEntitiesConfiguration;

@SpringBootApplication
@ComponentScan(basePackageClasses = { PmgrEntitiesConfiguration.class, Application.class })
@EnableTransactionManagement
public class Application extends SpringBootServletInitializer {

	public static void main(final String[] args) {
		SpringApplication.run(Application.class, args);
	}
}