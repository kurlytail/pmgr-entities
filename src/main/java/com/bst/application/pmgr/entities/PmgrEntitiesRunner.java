package com.bst.application.pmgr.entities;

import java.util.Random;

import javax.persistence.EntityManager;
import javax.transaction.Transactional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.CommandLineRunner;
import org.springframework.stereotype.Service;

import com.bst.pmgr.entities.Work;
import com.bst.pmgr.entities.repositories.WorkRepository;

@Service
@Transactional
public class PmgrEntitiesRunner implements CommandLineRunner {

	@Autowired
	private WorkRepository workRepository;

	@Autowired
	private EntityManager entityManager;

	@Override
	public void run(final String... args) throws Exception {

		final long countOfWork = Math.abs(new Random().nextLong()) % 10;
		long indexOfWork = 0;
		
		while (indexOfWork < countOfWork) {
			final Work work = new Work();
			this.workRepository.save(work);
			++indexOfWork;
		}

		this.entityManager.flush();
	}

}
