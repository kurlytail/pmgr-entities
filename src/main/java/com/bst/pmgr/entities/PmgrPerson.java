package com.bst.pmgr.entities;

import java.util.List;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.EntityListeners;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.OneToMany;
import javax.validation.constraints.NotNull;

import com.bst.pmgr.entities.audit.PmgrPersonAudit;
import com.bst.user.authentication.entities.Person;

@EntityListeners(PmgrPersonAudit.class)
@Entity
public class PmgrPerson {
	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	private Long id;

	@NotNull
	@Column(updatable = false)
	private Person person;

	@OneToMany(mappedBy = "createdBy")
	private List<Work> works;

	public Long getId() {
		return id;
	}

	public Person getPerson() {
		return person;
	}

	public void setId(Long id) {
		this.id = id;
	}

	public void setPerson(Person person) {
		this.person = person;
	}

	public List<Work> getWorks() {
		return works;
	}

}
