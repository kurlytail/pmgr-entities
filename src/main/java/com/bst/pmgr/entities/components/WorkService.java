package com.bst.pmgr.entities.components;

import org.springframework.stereotype.Service;

import com.bst.pmgr.entities.Work;

@Service
public class WorkService {
		
	public Work create(final String name) {
		return new Work();
	}

}
