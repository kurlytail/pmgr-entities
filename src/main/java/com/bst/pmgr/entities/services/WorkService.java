package com.bst.pmgr.entities.services;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.bst.pmgr.entities.PmgrPerson;
import com.bst.pmgr.entities.Work;
import com.bst.pmgr.entities.repositories.PmgrPersonRepository;
import com.bst.pmgr.entities.repositories.WorkRepository;
import com.bst.user.authentication.entities.Person;

@Service
public class WorkService {

	@Autowired
	private PmgrPersonRepository personRepository;

	@Autowired
	private WorkRepository workRepository;

	@Transactional
	public Work create(Person createdBy, String name) {
		PmgrPerson pmgrPerson = personRepository.findByPerson(createdBy);

		Work work = new Work();
		work.setName(name);

		work.setCreatedBy(pmgrPerson);
		return workRepository.save(work);
	}

}
