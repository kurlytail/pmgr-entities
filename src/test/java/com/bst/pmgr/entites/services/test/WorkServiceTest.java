package com.bst.pmgr.entites.services.test;

import static com.bst.utility.testlib.SnapshotListener.expect;

import javax.transaction.Transactional;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.domain.EntityScan;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.autoconfigure.orm.jpa.TestEntityManager;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.TestExecutionListeners;
import org.springframework.test.context.TestExecutionListeners.MergeMode;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.context.junit.jupiter.SpringExtension;

import com.bst.configuration.pmgr.entities.PmgrEntitiesConfiguration;
import com.bst.pmgr.entities.PmgrPerson;
import com.bst.pmgr.entities.Work;
import com.bst.pmgr.entities.services.WorkService;
import com.bst.user.authentication.entities.Person;
import com.bst.user.authentication.repositories.PersonRepository;
import com.bst.utility.testlib.SnapshotListener;

@ExtendWith(SpringExtension.class)
@DataJpaTest
@ContextConfiguration(classes = { PmgrEntitiesConfiguration.class })
@TestExecutionListeners(listeners = SnapshotListener.class, mergeMode = MergeMode.MERGE_WITH_DEFAULTS)
@EntityScan({ "com.bst.user.authentication.entities" })
@EnableJpaRepositories("com.bst.user.authentication.repositories")
@TestPropertySource("classpath:work-service-test.properties")
@Transactional
@ComponentScan({ "com.bst.pmgr.entities" })
class WorkServiceTest {

	@Autowired
	private TestEntityManager entityManager;
	
	@Autowired
	private PersonRepository personRepository;

	@Autowired
	private WorkService workService;

	@Test
	void testWorkCreate() throws Exception {
		Person person = new Person("jon@doe.com");
		person = personRepository.save(person);

		Work work = workService.create(person, "testWork");
		expect(work).toMatchSnapshot();
		
		entityManager.flush();
	}
	
	@Test
	void testWorkDelete() throws Exception {
		Person person = new Person("jon@doe.com");
		person = personRepository.save(person);

		Work work = workService.create(person, "testWork");
		entityManager.flush();
		
		work = entityManager.refresh(work);
		work.getCreatedBy().removeWork(work);
		entityManager.flush();
		
		PmgrPerson createdBy = entityManager.refresh(work.getCreatedBy());
		expect(createdBy).toMatchSnapshot();
		
		work = entityManager.find(Work.class, work.getId());
		expect(work).toMatchSnapshot();
	}
	
}
