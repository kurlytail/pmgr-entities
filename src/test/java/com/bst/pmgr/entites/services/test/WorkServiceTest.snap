{
  "testWorkCreate" : [ {
    "id" : 2,
    "name" : "testWork",
    "activities" : [ ],
    "documents" : [ ],
    "tools" : [ ],
    "processes" : [ ],
    "processGroups" : [ ]
  } ]
}